Added clear function to FrameBuffer.

sebftw opened the pull request:
Instead of calling window.clear() when we have a framebuffer activated we should call framebuffer.clear().
Then it makes sure to clear the renderbuffers that are set.
This also fixes: https://github.com/glumpy/glumpy/pull/189

rougier commented:
Good point. Maybe you could also add the color option (as for window clear), what do you think?

sebftw commented:
That is a good idea.

sebftw commented:
Should we then add a self._color attribute and color argument to the FrameBuffer class? Like with app.window.Window?

rougier commented:
Probably, yes.

sebftw commented:
I thought about it and removed the clear function from the framebuffer, but I now realize it is simple:
The window class clear function always does two things: It sets glClearColor and then calls glClear.
Hence **there is no way to mess up the window clear color**.

What I did now is that **I removed the other calls to glClearColor** in the window class. These were present **in on_init and color.setter**. They are redundant, since glClearColor will be called upon clearing anyways.

Should I just add the framebuffer clear function again?

rougier commented:
You're right. Window clear color is protected. For the `on_init`, we need to make sure `glClear` is not  called elsewhere (I mean, without using the `clear` function). 

For the framebuffer, we could probably do the same and have a dedicated `clear` function.

rougier commented:
I think I forgot about this PR, sorry. What did we decide in the end concerning the clear function?

sebftw commented:
No problem. I haven't had too much time to look into it anyways.
I think we decided to add a clear function to the FrameBuffer. I am unsure how/if the color attribute should be made for the FrameBuffer. It would be nice if it could be exactly the same as for the Window, but an issue is that the @color.setter is already defined, so we need to make a decision.
A uniform interface for clearing Windows and FrameBuffers or backwards compatibility (no color attribute)?


There are many ways to go about it and I must admit I am too inexperienced to make the call.
It is also possible to just add a setBackgroundColor function to both Window and FrameBuffer, bloating the code somewhat, but preserving backwards compatibility and giving a uniform interface.

rougier commented:
I think having a specific clear color for window and framebuffer makes sense. But as you noticed, _color is used in the framebuffer. I proposed to rename _color to _clearcolor that would be also similar to _clearflags in the window.py class. What do you think?

sebftw commented:
I agree. I'm thinking of adding a @clearcolor.setter to Window and FrameBuffer and just keeping the old @color.setter in them both for backwards compatibility.
Then changing any reference from color to clearcolor in the Window documentation, essentially depreciating it.
I will try implementing it when I get time. That is probably easier than explaining.
